### ISSUE TITLE
A first look at function builder

### ISSUE THREAD

**Issue by cboettig** [cboettig/build.eml#2]
@amoeba @mbjones Here's a first proof-of-concept on what an R package that defines a function for approximately each complex type would look like.  Curious what you think.  Fastest way to browse this is probably check out the pkgdown version of the docs here: https://cboettig.github.io/build.eml/reference/index.html, though I think the real test is in seeing if this is a more useful, more beginner-friendly way to construct EML than we can currently offer in EML.  I'm not sure.

As commented in https://github.com/ropensci/EML/issues/252, this is based on a crude parsing of EML XSD files which could definitely be improved upon to pull out the rest of the documentation and examples found there (though I really wish it were easier to do so! parsing the complex referential logic of XSD classes across multiple files really is rather messy to me).  Certainly this could be much improved (e.g., to start, adding cross-linking between the documentation files, and including the rest of the documentation strings indicating which value is required or not, etc).  

These functions are generated programmatically in a very simple way from this [hack-y table](https://github.com/cboettig/emld/blob/master/data-raw/eml-2.1.1.tsv) I [constructed by parsing](https://github.com/cboettig/emld/blob/master/data-raw/eml_schema_parser.R) the XSD files. It would be really useful to me to have a more complete and polished version of the XSD schema+documentation in a format like this or JSON that I could then work with, but not sure when I will get enough time to dive back into XSD extraction.   

As also anticipated, this approach creates a handful of namespace collisions with base packages, as noted in #1.  (with more collisions possible with other additional common packages).  Not sure how best to deal with that, but I'm not really sure these collisions are acceptable.

**Comment by mbjones** [comment]
@cboettig After a quick perusal, I think this looks awesome.  I haven't had time to try it out, but we could ask some folks on the data team at NCEAS that use the package daily for feedback.

Regarding parsing XSD files... are you building your own parser, or using an XML Schema [conformant parser](https://www.w3.org/TR/xmlschema-1/#concepts-conformance)?  Such a parser should manage all of the cross file imports and referencing seamlessly, and provide access to all of the schema components programatically. These are available in the [post-schema-validation infoset](https://www.w3.org/TR/xmlschema-1/#nsi-schema_components) (PSVI) from the parser (for example, the Java Xerces parser provides functions for retrieving all xml schema components through their [XML Schema API](https://www.w3.org/Submission/2004/SUBM-xmlschema-api-20040309/java-binding.html).  See the [XML Schema Abstract Model](https://www.w3.org/TR/xmlschema-1/#concepts-data-model) for the list of schema components, as well as the [XML schema UML diagram](https://www.w3.org/TR/xmlschema-1/#component-diagram) showing what schema components should be accessible from the parser. 

In terms of package naming conflicts, I suggest that prefixing all of the function names would help minimize conflicts (e.g., use `eml_title()` or `md_title()` rather than `title()`) and it would help group all of the related functions together in help indices.

**Comment by amoeba** [comment]
This is interesting!

Overall those docs look good but I think I'd want them to be more helpful with respect to guidance on what type of thing should be passed for each argument. For example, if I look at the [dataset](https://cboettig.github.io/build.eml/reference/dataset.html) page, the help for `alternateIdentifier` is just "A secondary identifier for this entity". I think this could be improved in two ways:

1. Give a little more guidance on how to fill this out, either in-line or as a link to a richer resource. This isn't actually a problem with `build.eml` but just our docs in general (I think). `build.eml` could simply expose that extra info if we take the time to author it somewhere.
2. Provide a type. For `alternateIdentifier`, we'd expect `character`, but, say, for `creator`, can I pass in a return value from `creator()` or a `person()` call?

Do you think (2) could be done automatically or would there be some gluing to do there?

**Comment by cboettig** [comment]
thanks @amoeba !  

Re the documentation detail, that's probably largely because we're just grabbing the `summary` section of the docs so far.  However, there are some places where the XSD documentation itself is too thin or possibly missing entirely

Re 2, yeah, we could probably improve on that.  Currently I've just updated the docs to attempt to link to the corresponding constructor (e.g. see https://cboettig.github.io/build.eml/reference/eml.html ) but I agree we could do more with typing.  

Both of these come down to mostly needing more parsing of the XSD.  @mbjones I'm just using the `xml2` R package, which is just binding `libxml2-dev` libraries, so all my parsing is just XPATH expressions.  Thanks much for these links, I'll dig in and see if they can help me.

**Comment by amoeba** [comment]
> Re the documentation detail,...

Makes sense. We've slowly found gaps in the in-schema docs but that's been piecemeal and there are definitely some big gaps remaining.

> Currently I've just updated the docs to attempt to link to the corresponding constructor 

:tada: nice!

**Comment by jeanetteclark** [comment]
This looks great! I have two comments:

1. For classes where the text value is duplicated (eg userID) it would be nice if the duplicated value was included as an argument in the function. This would enable you to do something like this:

```
jc_contact <- creator(individualName = 
                        individualName(givenName = "Jeanette", surName = "Clark"),
                      electronicMailAddress = "jclark@nceas.ucsb.edu", 
                      userId = userId(userId = "an orcid", directory = "https://orcid.org"))
```

as far as I can tell, currently you would have to add that in like this:

```
jc_contact$userId <- userId(directory = "https://orcid.org")
jc_contact$userId$userId <- "an orcid"
```

2. I would love to see good, valid examples in the documentation for each function. I think the datateam can help generate some of these if needed, they would be hugely beneficial to our team. Particularly with some of the more complex elements, you can spend a lot of time banging your head against the wall with validation errors as you are learning.
